universal/include/userver/formats/universal/common_containers.hpp

```diff
@@ -0,0 +1,368 @@
+  }
+};
+
+
+template <utils::ConstexprString Pattern>
+static const utils::regex kRegex(Pattern);
+
+struct Pattern : public impl::EmptyCheck, public impl::Param<const utils::regex*> {
+  constexpr inline Pattern(const utils::regex& regex) :
+      impl::Param<const utils::regex*>(&regex) {}
+  constexpr inline std::optional<std::string> Check(std::string_view str) const {
+    if(utils::regex_match(str, *this->value)) {
+      return std::nullopt;
+    }
+    return this->Error(str);
+  }
+  constexpr inline std::string Error(std::string_view) const {
+    return "Error";
+  }
+};
+
+struct AdditionalProperties : public impl::EmptyCheck, public impl::Param<bool> {
+  constexpr inline AdditionalProperties(const bool& value) :
+      impl::Param<bool>(value) {}
+};
+
+template <typename T>
+struct FieldConfig<std::optional<T>> {
+  FieldConfig<T> Main = {};
+  std::optional<Default<T>> Default = std::nullopt;
+  bool Required = false;
+  bool Nullable = false;
+  static inline constexpr auto Checker(const auto& check, const auto& value) -> std::optional<std::string> 
+      requires requires{check.Check(*value);} {
+    if(value) {
+      return check.Check(*value);
+    };
+    return std::nullopt;
+  }; 
+  constexpr inline std::optional<std::string> Check(const std::optional<T>& value) const {
+    if(value) {
+      return this->Main.Check(*value);
+    };
+    return std::nullopt;
+  }
+  constexpr auto Write(const std::optional<T>& value, std::string_view field_name, const auto& names, auto& builder) const {
+    if(value) {
+      this->Main.Write(*value, field_name, names, builder);
+      return;
+    }
+    if(this->Default) {
+      this->Main.Write(this->Default->value(), field_name, names, builder);
+      return;
+    }
+  }
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<T> Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    if(this->Required && value[name].IsMissing()) {
+      return std::nullopt;
+    }
+    if(!value[name].IsMissing()) {
+      return this->Nullable 
+      ? (value[name].IsNull()
+        ? std::nullopt
+        : std::optional{this->Main.template Read<MainClass, I>(std::forward<Value>(value))})
+      : std::optional{this->Main.template Read<MainClass, I>(std::forward<Value>(value))};
+    }
+    if(this->Default) {
+      return this->Default->value();
+    }
+    return std::nullopt;
+  }
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<std::optional<T>> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    if(value[name].IsMissing() && this->Required) {
+      return std::nullopt;
+    }
+    if(this->Nullable) {
+      std::optional response = this->Main.template TryRead<MainClass, I>(std::forward<Value>(value));
+      if(response) {
+        return response;
+      }
```

Comment:
`Default` should come into play when `value` is null or missing (and the null-ness type is supported according to flags), not when the `value` is present and its parsing failed

So:
- when parsing returns a non-empty optional, wrap it in another optional
- when parsing returns an empty optional, return top-level `std::nullopt`